src/vm/eventpipe.h

```diff
@@ -63,7 +63,7 @@ class EventPipe
     static bool Enabled()
     {
         LIMITED_METHOD_CONTRACT;
-        return s_tracingInitialized && (s_numberOfSessions > 0);
+        return s_tracingInitialized && (s_numberOfSessions.LoadWithoutBarrier() > 0);
```

Comment:
nit: AFAIK `s_numberOfSessions` is incremented/decremented under a lock with `++` and `--` operator, and that implicitly calls `Load()` inside it and can hit the memory barrier (and therefore, perf). Do we want to change those to `LoadWithoutBarrier()` followed by `Store()`? FWIW I don't think it's a huge deal since this code path isn't called that often (unless someone abuses it) but maybe it's something we want to think about :/

Reply:
Although you could change those locations I wouldn't : )

To pull back the curtain a bit so it doesn't seem like these suggestions come out of thin air... For every piece of code in the runtime it gets used in some different scenarios, and then each of those scenarios has a perf goal. We could certainly do a better job at being more explicit documenting and sharing what the perf goals are, but for this particular case, I think these are reasonable goals:

a) Enable/Disable a session - a millisecond (not counting rundown)
b) Check if event is enabled - 10s of CPU cycles

Coming up with the goals is probably the hard part and where it is easy for people to reach different conclusions, but mostly this is just a matter of thinking about what customers want to do with the feature and what might be acceptable to them. In the first case turning on/off sessions seems like something that happens at human timescales (someone runs dotnet-trace), or for EventListener they probably turn it on when an app starts up and then turn it off much later (seconds to hours later). Most humans have a hard time discerning time differences below 0.01 sec, and in many cases of UI update anything around 0.1 feels responsive. I picked 1 ms because if we are doing it on startup we care about that in aggregate and milliseconds for different things add up. For the second one, this check is something that gets invoked all over the runtime in performance critical paths like GC. There could be hundreds of potential logging points and when logging is off we expect virtually no overhead on operations that were already being measured in microseconds in order to meet their performance goals. In order to register near zero overhead our check needs to be 3+ orders of magnitude faster than that, ie nanoseconds.


Once you have the goal as your baseline, then we can ask how the thing we are thinking about optimizing compares to the overall goal. In the case of a memory barrier, I think of it being in the vicinity of 30-100 cycle delay. I'm not certain that is terribly accurate, but its probably in the right order of magnitude. If the goal is 10s of cycles, the delay is huge, potentially larger than the entire performance budget for this opation. If the goal is 3MM cycles then it only represents 0.001%. Once I am looking at something less than 1% of my perf goal generally there are other aspects that are more important, such as readability and maintainability of the code. At the very least such optimizations feel 'premature' in the classic Donald Knuth sense and would be better directed by top-down performance measurement. Thats why I wind up suggesting LoadWithoutBarrier() at this callsite, but not at any of those other callsites.

Hope that helps : )

Reply:
Thanks for the clarification Noah! Always something new to learn from you... : )